### ISSUE TITLE
Explore BlackOut approach to speed up SoftMax for large classification output

### ISSUE THREAD

**Issue by nyghtowl** [eclipse/deeplearning4j#1491]
Reference paper for enhancement http://arxiv.org/pdf/1511.06909v7.pdf

This approach has been used in problems like ImageNet when output classes are 22K or more
Use subset of output layer vs entire output (e.g. subsampling)

**Comment by dapifer** [comment]
Planning to do this.  Based on the paper I believe this will require #3005 to be done, and then Blackout should be able to be entirely a set of extensions.  Specifically (and forgive the refs to the paper, but if we get sidetracked hopefully it makes it easier for someone else to pick up):

BlackoutLossFunction implements ILossFunction, to compute the score and gradient based on the paper.  ComputeScore boils down to eqn 6.  ComputeGradiant amounts to eqn 10.  The plan is these will get passed in the subsampled preOutput and a BlackoutActivation.

BlackoutActivation implements IActivation, to compute tilde-p which is a subsampled softmax over the subsampled preOutput.  BlackoutActivation will (unless there's a better way to do this!) have a "setIterationQValues" function so that the element selection probabilities q_i in equation 7 can be supplied by the BlackoutLayer.

BlackoutLayer ... I think extends OutputLayer? ... to do some special things during forward activation.
* Statically (once per model) precompute a CDF for the subselection probabilities so we can efficiently sample output indexes.
* Generate a random index set Sk (once per forward activation -- i.e. each instance of the output layer in an RNN will need to do this).
* Provide the subselection probabilities corresponding to the elements of Sk to the activation function (assuming the activation function gets a single instance per output layer instance, in an unrolled RNN).
* Provide the subselection probabilities and the subset indexes Sk to the updater.

BlackoutUpdater implements the as-yet-nonexistant IUpdater, and will take Sk indexes, q values, and gradient to apply eqns 17-19 to the weight set.

**Comment by AlexDBlack** [comment]
@dapifer FYI - some recent progress on this issue - https://github.com/deeplearning4j/deeplearning4j/pull/3373

I need to dig into the blackout paper and think about it a bit before replying re: your proposed design.

**Comment by dapifer** [comment]
You're going to make me look bad, I'd better find some time for this, later
in the week.  One of my coworkers was looking at the DL4J architecture a
bit more and we will probably start with something that doesn't break
out the loss function or activation function from the output layer.  The
updater, we can't see a way around, and we're looking at whether we need a
custom Solver or StepFunction, because we might.  To be continued...

**Comment by AlexDBlack** [comment]
Updater optimisations are now merged: https://github.com/deeplearning4j/deeplearning4j/pull/3373
There is a unit test in there with an example of using custom updaters.
I still haven't got to looking at the blackout paper in any detail.

**Comment by agibsonccc** [comment]
Closing stale issue: cleanup.
